Use cryptography and pyOpenSSL instead of M2Crypto

jeremycline opened the pull request:
M2Crypto is Python 2 only and is in maintenance mode. Although pyOpenSSL
is _also_ in maintenance mode, it supports Python 3 and relies on
cryptography for its OpenSSL bindings. Currently the only thing
cryptography is missing (that pyOpenSSL has) is certificate chain
validation. Once that's available in cryptography, we can drop
pyOpenSSL.

Unfortunately, we still need to support M2Crypto because the current
versions of cryptography and pyOpenSSL in EL7 is below 1.6 and 16.1.0
respectively.

This patch uses cryptography and pyOpenSSL if available and falls back
to M2Crypto.

fixes #411

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Use cryptography and pyOpenSSL instead of M2Crypto

M2Crypto is Python 2 only and is in maintenance mode. Although pyOpenSSL
is _also_ in maintenance mode, it supports Python 3 and relies on
cryptography for its OpenSSL bindings. Currently the only thing
cryptography is missing (that pyOpenSSL has) is certificate chain
validation. Once that's available in cryptography, we can drop
pyOpenSSL.

Unfortunately, we still need to support M2Crypto because the current
versions of cryptography and pyOpenSSL in EL7 is below 1.6 and 16.1.0
respectively.

This patch uses cryptography and pyOpenSSL if available and falls back
to M2Crypto.

fixes #411

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+        signer (str): The Common Name of the certificate used to sign the message.
+
+    Returns:
+        bool: True of the policy defined in the settings allows the signer to send
```
True *if* the policy…

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+            # whitelisted to send on this topic in our config policy.
+            return True
+        else:
+            # We have a policy for this topic and $homeboy isn't on the list.
```
Haha

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+            # routing_policy for the topic of this message.. but we don't
+            # really care.
+            _log.info('No routing policy defined for "{t}" and routing_nitpicky '
+                      'is False so the message is being treated as authorized.'.format(t=topic))
```
This might make sense at warning level.

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+def load_remote_cert(location, cache, cache_expiry, tries=0, **config):
+    """ Get a fresh copy from fp.o/fedmsg/crl.pem if ours is getting stale.
+
+    Return the local filename.
```
I recommend documenting the arguments and the return type.

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+        (cache_expiry and time.time() - modtime > cache_expiry)
+    ):
+        try:
+            response = requests.get(location)
```
Maybe put a timeout on this call to ```get()```.

bowlofeggs commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,104 @@
+                os.makedirs(usr_dir)
+
+            with open(cache, 'w') as f:
+                f.write(response.content)
```
Note that this could also raise an IOError, but that's probably OK.

bowlofeggs commented on fedmsg/crypto/x509.py:
```diff
@@ -196,87 +196,18 @@ def fail(reason):
-        (cache_expiry and time.time() - modtime > cache_expiry)
-    ):
-        try:
-            response = requests.get(location)
```
Oh I see - you just moved this code block. I'd still recommend that timeout though… (maybe a separate PR or commit if you prefer - up to you).

jeremycline replied:
I was actually thinking of doing a separate PR that changes this API a lot (one reason I didn't bother documenting it when I moved it). It does quite a few surprising things which I will probably address in the fix for #420.

bowlofeggs commented on fedmsg/crypto/x509_ng.py:
```diff
@@ -19,256 +19,215 @@
-    except M2Crypto.RSA.RSAError as e:
-        return fail(str(e))
+        _validate_signing_cert(ca_certificate, certificate, crl)
+    except X509StoreContextError as e:
```
I presume this step will catch expired certificates and non-properly signed certificates?

jeremycline replied:
Yup, it'll raise for that as well as revoked certificates and expired CRLs

bowlofeggs commented on fedmsg/tests/crypto/test_x509.py:
```diff
@@ -0,0 +1,216 @@
+# This file is part of fedmsg.
+# Copyright (C) 2012 - 2014 Red Hat, Inc.
```
2017?

bowlofeggs commented on fedmsg/tests/crypto/test_x509.py:
```diff
@@ -0,0 +1,216 @@
+        self.config['routing_policy'] = {'mytopic': ['shell-app01.phx2.fedoraproject.org']}
+        self.config['routing_nitpicky'] = True
+        signed = self.sign({'topic': 'mytopic', 'message': 'so secure'}, **self.config)
+        self.assertTrue(self.validate(signed, **self.config))
```
How about a ```test_signed_by_expired_cert``` and ```test_signed_by_expired_ca``` (bad name, but what I mean is the signing cert is signed by an expired CA) as well?

jeremycline replied:
Good idea! I have made it so, Number One.

bowlofeggs commented on fedmsg/tests/crypto/test_x509.py:
```diff
@@ -0,0 +1,216 @@
+
+
+@skipIf(not (_m2crypto or _cryptography), "Neither M2Crypto nor Cryptography available")
+class X509BaseTests(TestCase):
```
I like this pattern of making a superclass to get the same tests run with the different implementations. Nice!

bowlofeggs commented on fedmsg/tests/crypto/test_x509.py:
```diff
@@ -0,0 +1,216 @@
+            'ca_cert_cache': os.path.join(SSLDIR, 'ca.crt'),
+            'ca_cert_cache_expiry': 1497618475,  # Stop fedmsg overwriting my CA, See Issue 420
+
+            'crl_location': "http://threebean.org/fedmsg-tests/crl.pem",
```
Do these tests work offline? If not, I see that you have a ```fedmsg/tests/test_certs/keys/crl.pem``` that you might could use instead…

jeremycline replied:
They do, but only because I set the cache expiry to be ~47 years. There's no way to have the cert not downloaded from the Internet eventually and it defaults to the Fedora CA and CRL so even if you don't provide a URL it'll happen (all of which I want to change, I just haven't decided how yet).

bowlofeggs reviewed (approved):
This looks reasonable to me. Note however that I lack familiarity with the cryptography and pyopenssl libraries, so I am not able to comment on whether they were used correctly or not. Thus, I recommend getting a second opinion from someone who is familiar with those APIs before merging. On the other hand, I did like the tests you wrote and they helped give me confidence in the code ☺

jeremycline committed:
Start linting the crypto package

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Add docblock load_remote_cert

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Enable CRLs to be unspecified

Prior to this commit, it was required that a CRL was specified. This was
fine with the old M2Crypto code that didn't properly check the CRL for
expiration, but the new crytography-based code will fail if the CRL
is expired. This allows the definition of a CRL in the configuration to
be optional so users don't have to keep a fresh CRL that they may not
actually be using available.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Add a test for signatures by expired CAs

Thanks bowlofeggs for the suggestion!

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

puiterwijk commented on fedmsg/crypto/utils.py:
```diff
@@ -0,0 +1,113 @@
+
+    Return the local filename.
+
+    .. note:: This is not a public API and is subject to change.
```
Maybe start the name with _ then?

puiterwijk commented on fedmsg/crypto/x509.py:
```diff
@@ -19,46 +19,50 @@
+def _disabled_sign(*args, **kwargs):
+    """A fallback function that emits a warning when no crypto is being used."""
+    warnings.warn('Message signing is disabled because "cryptography" and '
+                  '"pyopenssl" or "m2crypto" are not available.')
```
Perhaps add parentheses to indicate the and/or relations?

puiterwijk commented on fedmsg/crypto/x509_ng.py:
```diff
@@ -19,256 +19,221 @@
+        fedmsg.encoding.dumps(message).encode('utf-8'),
         asymmetric.padding.PKCS1v15(),
-        hashes.SHA1()
+        hashes.SHA1(),
```
While it's fine right now, it would be great if we could start migrating to new hashes, like sha256/sha512.
Maybe add an indicator which is used before the next version, so we can switch easier in the future?

jeremycline replied:
I considered this, but I didn't want to make that change alongside all the other stuff that's already happening. I thought I filed an issue, but it seems like I hadn't, so I filed #438

puiterwijk replied:
That makes perfect sense. This was just an idea now that you got me rereading the code :).

puiterwijk commented on fedmsg/crypto/x509_ng.py:
```diff
@@ -19,256 +19,221 @@
-    # Load our policy from the config dict.
+    # Step 4, check that the certificate is permitted to emit messages for the
+    # topic.
+    common_name = crypto_certificate.subject.get_attributes_for_oid(x509.oid.NameOID.COMMON_NAME)
```
You should not use the common name ideally, this is being deprecated by the CA/Browser Forum baseline requirements.
While we are not reliant on those, I think ideally we should follow their move to the Subject Alternative Name.
This does mean there may be multiple records, and we might need to check all SAN values in the authorization step.
For the time being, for compatibility, we might add the Common Name to the list of SANs for checking purposes though and issue a warning if the CN was not in any SAN values.

jeremycline replied:
Yeah, I wanted to do this, but the M2Crypto version is only using the CN. I will definitely look into moving both of them to SAN (#439).

puiterwijk replied:
The concrete sentence about Common Name from the CA/Browser Baselines requirements v1.4.5 (which is currently the latest stable, source https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.4.5.pdf):

````
Certificate Field: subject:commonName (OID 2.5.4.3)   
Required/Optional: Deprecated (Discouraged, but not prohibited) 
````

puiterwijk commented on fedmsg/crypto/x509_ng.py:
```diff
@@ -19,256 +19,221 @@
+    common_name = common_name[0]
     routing_policy = config.get('routing_policy', {})
+    nitpicky = config.get('routing_nitpicky', False)
+    return utils.validate_policy(
```
An idea would be a return any(...) over all the SANs.

puiterwijk commented on fedmsg/tests/crypto/test_x509.py:
```diff
@@ -0,0 +1,237 @@
+            'ca_cert_cache': os.path.join(SSLDIR, 'ca.crt'),
+            'ca_cert_cache_expiry': 1497618475,  # Stop fedmsg overwriting my CA, See Issue 420
+
+            'crl_location': "http://threebean.org/fedmsg-tests/crl.pem",
```
Ideally we would not depend on this website... Maybe use a file:// ?

jeremycline replied:
Most unfortunately, the way the code currently works it doesn't support file://. I set the cache expiry to be sometime in the mid 2030s so it should never actually try to fetch from that URL. I plan to re-write all the cert fetching code in the near future.

puiterwijk reviewed (approved):
Overall this looks good, with some future suggestions.

jeremycline committed:
Mark load_remote_cert as private

Signed-off-by: Jeremy Cline <jeremy@jcline.org>
